engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/MetadataStorage.java

```diff
@@ -0,0 +1,214 @@
+    return metadata.entrySet().stream().map(
+      (en) -> fn.apply(en.getKey(), en.getValue())
+    ).toList();
+  }
+
+  /** Prepares the metadata for serialization.
+    *
+    * This operation takes place _in place_.
+    *
+    * Metadata prepared for serialization should not contain any links that
+    * span more than one module, or any other properties that are problematic
+    * when serialized.
+    *
+    * Due to the type safety properties of
+    * [[org.enso.compiler.core.ir.MetadataStorage]], to allow this conversion
+    * to work it must be type-refined to return `typeof this`. To that end,
+    * there is no default definition for this method.
+    *
+    * @param compiler the Enso compiler
+    */
+  public final void prepareForSerialization(CompilerStub compiler) {
+    var newMap = metadata.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, (en) -> {
+      var value = en.getValue();
+      var newVal = value.prepareForSerialization(compiler);
+      return newVal;
+    }));
+    this.metadata.putAll(newMap);
+  }
+
+  /** Restores metadata after it has been deserialized.
+    *
+    * Due to the type safety properties of
+    * [[org.enso.compiler.core.ir.MetadataStorage]], to allow this conversion
+    * to work it must be type-refined to return `typeof this`. To that end,
+    * there is no default definition for this method.
+    *
+    * @param compiler the Enso compiler
+    * @return `true` if restoration was successful, `false` otherwise
+    */
+  public boolean restoreFromSerialization(CompilerStub compiler) {
+    var ok = new boolean[] { true };
+    var newMap = metadata.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, (en) -> {
+      var value = en.getValue();
+      var newOption = value.restoreFromSerialization(compiler);
+      if (newOption.nonEmpty()) {
+        return newOption.get();
+      } else {
+        ok[0] = false;
+        return value;
+      }
+    }));
+    this.metadata = newMap;
+    return ok[0];
+  }
+
+  @Override
+  public String toString() {
+    var sb = new StringBuilder("MetadataStorage[");
+    var names = new ArrayList<String>();
+    for (var v : metadata.values()) {
+      var m = (ProcessingPass.Metadata) v;
+      names.add(m.metadataName());
+    }
+    Collections.sort(names);
+    var sep = "";
+    for (var n : names) {
+      sb.append(sep);
+      sb.append(n);
+      sep = ", ";
+    }
+    sb.append("]");
+    return sb.toString();
+  }
+
+  private static final Comparator<ProcessingPass> COMPARATOR = (p1, p2) -> {
+    return p1.getClass().getName().compareTo(p2.getClass().getName());
+  };
+
+  private Map<ProcessingPass, ProcessingPass.Metadata> copyMetaMap() {
+    var copy = new TreeMap<ProcessingPass, ProcessingPass.Metadata>(COMPARATOR);
+    copy.putAll(metadata);
+    return copy;
+  }
+
+  @Override
+  public int hashCode() {
+    int hash = 5;
+    hash = 17 * hash + Objects.hashCode(this.metadata);
```

Comment:
What is the reason for this custom hashCode? I have seen this somewhere else as well. Why not only delegate to `Objects.hashCode(this.metadata)`?